### PR TITLE
Add new filter type v0 for segwit only Scripts to blockfilterindex

### LINKED ISSUE #18222: Add `blockfilterindex = p2wpkh`
I would like to have a `p2wpkh` `blockfilterindex` type to outsource Wasabi's filter generation to Bitcoin Core, as our filter generation is orders of magnitude slower and we would like to bring this to the client side too, so our users with full nodes would be able to validate that our server does not cheat.  

I previously created a `blockfilterindex` extensibility proposal to enable filter type combinations: https://github.com/bitcoin/bitcoin/issues/18221. So the followup PR of this issue will make sure it adheres to it. Since the PR is ready already, I'll omit to go into the implementation details in this issue, rather concentrate on the goal, the benefits of end users and how we ruled out the alternative solutions.

### The Goal

We would like to enable our users to upgrade Wasabi to be a full node with one click.

image

This feature is already implemented and rolled out to production (shipping with Core, fee estimation, tx broadcasting, mempool serving etc..) except the (most important) validation part of it is yet to be added.

### Alternative Solutions

I've been long struggling with how to enable this for our users and we ended up working together with a brilliant dev, @dangershony, who maintained the C# Bitcoin full node for years. Unfortunately we had to rule out all the possible alternatives.  

One alternative would be using `importmulti`, but we would not like to rely on the wallet feature of Bitcoin Core, rather stick to the node functionality.

Another alternative could be to slowly replace our filters with Bitcoin Core's filters on the client side, but that would require reindexing all of our wallets, which in some cases could take half a day. Also mixing filter types feels dirty and can result in unexpected issues.

We also considered bringing our own filter generation algorithm to the client side, but this is one of the oldest brain class of our codebase that serves a critical functionality so touching this could be problematic. It currently takes 2-3 days for us to create our filters on a powerful server, even though we are using `p2wpkh` filters and only building them from segwit activation. Bitcoin Core creates these filters within 2 hours. Why we are so slow is out of the scope of this issue, but here's the clue: input scripts aren't part of a transaction.

Further we considered just blindly going through all the blocks and process all the transactions and make sure the server doesn't lie. The performance here is also unacceptable.

Finally we would also don't like to use bloom filters, as it is on the edge of obsolation and having multiple filters to achieve the same purpose feels dirty from a complexity point of view.

### Benefits to the End User

@gmaxwell has suggested to articulate the benefits to the end user for changes to the codebase, thus this section.

Adding new filter types and filter type combinations (https://github.com/bitcoin/bitcoin/issues/18221) would enable a privacy oriented light wallet surge to be built on top of Bitcoin Core.

Maybe more importantly Bitcoin Core providing the filters to Wasabi enables a clean architecture and a seamless UX for users willing to run a full node that comes with Wasabi.

It is quite unique in a sense that no other popular light wallets went to the extent of building a full node into their software, especially not in a way that only requires a a single click configuration, so if our feature gets out of "work in progress" phase there may be an influx of full node users, which is good for the network.

What we stand to lose is that this full node feature will be delayed into the future so far that I may not be here when the next clean solution presents itself and maybe the Wasabi team won't be willing to work on it as frankly it has little business value.

### What's Next?

This issue described our motivation and the context of the feature request, our pull request is coming up soon, so we discuss the specifics.

### PR DESCRIPTION
**Pull request by dangershony**
**Introduce a new BlockFilter type [v0] next to the existing basic filter type, which is described in BIP 158.**
https://github.com/bitcoin/bips/blob/master/bip-0158.mediawiki#block-filter

Resolves #18222 

The BlockFilterIndex implementation allows for the addition of other filter types, this PR will add a new filter type for segwit script programs v0  types only.
Light clients (especially new clients that deal with segwit addresses only) do not need any legacy script types in the filters, and with this PR can choose to only index a much smaller portion of the chain.
That will greatly reduce the amount of data on disk and that is fetch over the network.

The same GCS parameters where used as basic type.

Total size of serialized filers as of block 619 361:
- Basic = 4.76 GB
- v0 = 252 MB 

Building the Index time
- Basic = 3.5 hours
- v0  = 3.5 hours 

Enabled in the config by:
```
blockfilterindex=p2wpkh
```
Enabling more then one filter:
```
blockfilterindex=v0
blockfilterindex=basic
```

### CODE REVIEW

**Review by nopara73 — CHANGES_REQUESTED** [review]
Can you add some tests?

File: src/blockfilter.h [hunk 1]
```diff
@@ -88,6 +88,8 @@ constexpr uint32_t BASIC_FILTER_M = 784931;
 enum class BlockFilterType : uint8_t
 {
     BASIC = 0,
+    // Floter 1 is reserved as option to include all filters
```
Comment (molnard) [inline src/blockfilter.h]:
```suggestion
    // Filter 1 is reserved as an option to include all filters.
```

File: src/blockfilter.cpp [hunk 2]
```diff
@@ -309,6 +347,13 @@ bool BlockFilter::BuildParams(GCSFilter::Params& params) const
         params.m_P = BASIC_FILTER_P;
         params.m_M = BASIC_FILTER_M;
         return true;
+    case BlockFilterType::P2WPKH:
+        params.m_siphash_k0 = m_block_hash.GetUint64(0);
+        params.m_siphash_k1 = m_block_hash.GetUint64(1);
+        // using the same filter params as basic type
```
Comment (molnard) [inline src/blockfilter.cpp]:
```suggestion
        // Using the same filter params as basic type.
```

**Review by molnard — CHANGES_REQUESTED** [review]

**Review by molnard — CHANGES_REQUESTED** [review]
PR name: filer => filter
Add new filer type p2wpkh to blockfilterindex
Add new filter type p2wpkh to blockfilterindex

**Comment by nopara73** [comment]
Note this PR and idea conforms to the original BIP157 filter extensibility proposal (https://github.com/bitcoin/bips/blob/master/bip-0157.mediawiki#filter-types) and to the proposal I submitted about filter combinations too: https://github.com/bitcoin/bitcoin/issues/18221

**Comment by nopara73** [comment]
For the record there is a discussion about that probably it should combine both P2WPKH and P2WSH scripts: https://github.com/bitcoin/bitcoin/issues/18222#issuecomment-592631517

**Review by nopara73 — APPROVED** [review]
tACK. Code LGTM. SegWit v0 filters' size: `151MB`.

**Review by Sjors — COMMENTED** [review]
Travis is failing do to a whitespace linter issue.

I left some code review comments; it's generally not advisable to spend too much time improving code before a concept ACK. I'm happy to review, as I'm inclined to concept ACK this, but we don't want to be stuck with an abandoned filter type (even though it looks pretty low maintance).

File: src/blockfilter.cpp [hunk 3]
```diff
@@ -309,6 +349,13 @@ bool BlockFilter::BuildParams(GCSFilter::Params& params) const
         params.m_P = BASIC_FILTER_P;
         params.m_M = BASIC_FILTER_M;
         return true;
+    case BlockFilterType::V0:
```
[the remark this answers is not in the document]
Comment (dangershony) [reply src/blockfilter.cpp]: ah good point, will fix.

File: src/blockfilter.cpp [hunk 4]
```diff
@@ -297,7 +329,15 @@ BlockFilter::BlockFilter(BlockFilterType filter_type, const CBlock& block, const
...
-    m_filter = GCSFilter(params, BasicFilterElements(block, block_undo));
+
+    switch (m_filter_type) {
+    case BlockFilterType::BASIC:
+        m_filter = GCSFilter(params, BasicFilterElements(block, block_undo));
+        break;
+    case BlockFilterType::V0:
+        m_filter = GCSFilter(params, V0FilterElements(block, block_undo));
+        break;
+    }
```
[the remark this answers is not in the document]
Comment (dangershony) [reply src/blockfilter.cpp]: It's handled in the code above so I left it out, but for best practice I agree so I'll add an assert as suggested.

File: src/blockfilter.cpp [hunk 5]
```diff
@@ -279,6 +281,36 @@ static GCSFilter::ElementSet BasicFilterElements(const CBlock& block,
...
+        }
+    }
+
+    for (const CTxUndo& tx_undo : block_undo.vtxundo) {
+        for (const Coin& prevout : tx_undo.vprevout) {
+            const CScript& scriptPubKey = prevout.out.scriptPubKey;
+            int witnessversion;
+            std::vector<unsigned char> witnessprogram;
+            if (scriptPubKey.empty() || !scriptPubKey.IsWitnessProgram(witnessversion, witnessprogram)) continue;
+            if (!(witnessversion == 0 && (witnessprogram.size() == WITNESS_V0_KEYHASH_SIZE || witnessprogram.size() == WITNESS_V0_SCRIPTHASH_SIZE))) continue;
```
[the remark this answers is not in the document]
Comment (dangershony) [reply src/blockfilter.cpp]: Is `opreturn` not already filtered out with the witness check?
Or you mean to add the `opreturn` check to the blockundo on the code of the `basic` filter type?

File: src/blockfilter.cpp [hunk 6]
```diff
@@ -279,6 +281,36 @@ static GCSFilter::ElementSet BasicFilterElements(const CBlock& block,
...
+{
+    GCSFilter::ElementSet elements;
+
+    for (const CTransactionRef& tx : block.vtx) {
+        for (const CTxOut& txout : tx->vout) {
+            const CScript& scriptPubKey = txout.scriptPubKey;
+            int witnessversion;
+            std::vector<unsigned char> witnessprogram;
+            if (scriptPubKey.empty() || !scriptPubKey.IsWitnessProgram(witnessversion, witnessprogram)) continue;
+            if (!(witnessversion == 0 && (witnessprogram.size() == WITNESS_V0_KEYHASH_SIZE || witnessprogram.size() == WITNESS_V0_SCRIPTHASH_SIZE)))  continue;
```
[the remark this answers is not in the document]
Comment (dangershony) [reply src/blockfilter.cpp]: In that case the method name should change to `BuildFilterElements` no?

File: src/blockfilter.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (Sjors) [reply src/blockfilter.cpp]: If you combine this with the code from the `basic` filter then it's probably more readable to `OP_RETURN` back for that as well. For the `v0` filter it indeed doesn't matter.

**Comment by Sjors** [comment]
You may want to squash all this into 1 commit (perhaps a separate commit to add `OP_RETURN` to the `basic` undo filter).

File: src/blockfilter.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (dangershony) [reply src/blockfilter.cpp]: shall I add the `OP_RETURN ` check also on the spent script then? I see no reason why not to check there too.

File: src/blockfilter.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (dangershony) [reply src/blockfilter.cpp]: Any particular reason why defaulting segwit to true?

File: src/blockfilter.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (dangershony) [reply src/blockfilter.cpp]: Done

File: src/blockfilter.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (dangershony) [reply src/blockfilter.cpp]: Done

[the remark this answers is not in the document]
Comment (Sjors) [reply src/blockfilter.cpp]: No, it should be `false` I think (basic filter is the default)

**Comment by dangershony** [comment]
Thank you @Sjors for your helpful review, I will squash the commits when there is a general ack.

**Comment by MarcoFalke** [comment]
Tend to NACK. I fail to see a use case for this, and a missing use case does not justify the time spent on code review and the resulting complexity/user confusion and maintenance overhead when this is merged.

**Comment by nopara73** [comment]
What do you mean no use case? It's already being used and in production for 1.5 years now.

**Comment by luke-jr** [comment]
Concept NACK if I understand this correctly.

If we assume everything will be Segwit eventually, this filter will just become redundant with the basic one?

If the goal is to reduce the initial download, couldn't the wallet just start with its birthdate block anyway?

Plus, remotely-accessible filters are still a bad idea in the first place, which this seems to presuppose...

File: src/blockfilter.cpp [hunk 7]
```diff
@@ -297,12 +315,22 @@ BlockFilter::BlockFilter(BlockFilterType filter_type, const CBlock& block, const
...
-    m_filter = GCSFilter(params, BasicFilterElements(block, block_undo));
+
+    switch (m_filter_type) {
+    case BlockFilterType::BASIC:
+        m_filter = GCSFilter(params, BuildFilterElements(block, block_undo));
+        break;
+    case BlockFilterType::V0:
+        m_filter = GCSFilter(params, BuildFilterElements(block, block_undo, true));
+        break;
+    default: assert(false);
```
Comment (luke-jr) [inline src/blockfilter.cpp]: Please use `case BlockFilterType::INVALID` here so new enum values warn at compile time.

File: src/blockfilter.cpp [hunk 8]
```diff
@@ -297,12 +315,22 @@ BlockFilter::BlockFilter(BlockFilterType filter_type, const CBlock& block, const
...
+    }
 }
 
 bool BlockFilter::BuildParams(GCSFilter::Params& params) const
 {
     switch (m_filter_type) {
+    case BlockFilterType::V0:
     case BlockFilterType::BASIC:
```
Comment (luke-jr) [inline src/blockfilter.cpp]: Let's keep the order BASIC,V0 here?

File: src/blockfilter.cpp [hunk 9]
```diff
@@ -257,60 +257,46 @@ const std::string& ListBlockFilterTypes()
...
+            }
             elements.emplace(script.begin(), script.end());
         }
     }
 
     for (const CTxUndo& tx_undo : block_undo.vtxundo) {
         for (const Coin& prevout : tx_undo.vprevout) {
             const CScript& script = prevout.out.scriptPubKey;
-            if (script.empty()) continue;
+            if (script.empty() || script[0] == OP_RETURN) continue;
```
[the remark this answers is not in the document]
Comment (luke-jr) [reply src/blockfilter.cpp]: Actually, it's it absolutely impossible to encounter an `OP_RETURN` here? These are spent pubkeys, right? You can't spend an `OP_RETURN`...

[the remark this answers is not in the document]
Comment (dangershony) [reply src/blockfilter.cpp]: I agree but this is in response to this comment (for readability)
https://github.com/bitcoin/bitcoin/pull/18223#discussion_r386120920

[the remark this answers is not in the document]
Comment (luke-jr) [reply src/blockfilter.cpp]: I don't think this makes it any more readable...

Prefer a comment simply stating that it is impossible.

[the remark this answers is not in the document]
Comment (dangershony) [reply src/blockfilter.cpp]: Reverted.

**Comment by nopara73** [comment]
@dangershony Let's discuss alternative solutions. If I wouldn't be so biased I think even I would NACK this with P2P in mind. (Not for RPC only though.)

**Comment by dangershony** [comment]
After internal discussions we decided to close this PR (thanks for providing review)

**Comment by MaxHillebrand** [comment]
Thanks for the review everyone!

Though I am sad about the concept NACK. I had high hopes that with customizable block filters served over RPC, we make a great step towards further separating the wallet functionality from Core, making it more compatible with dedicated wallet software. Wasabi is currently segwit v0 only, and it will be segwit v1 taproot only as soon as possible [similar to lightning wallets too]. Thus the custom filters have a superb use case, as I see it. Basic filters are unfortunately not feasible for Wasabi [even after segwit activation, basic filters are 4.8 GB; compared to 250MB for segwit v0 filters]. I do not know of another way to provide full verification while preserving the "light client feel" of Wasabi. This means, very unfortunately, that Wasabi will not have full verification for the foreseeable future, and the current Bitcoin Core integration is only a gimmicky bandwidth savings, not much more.

Anyway, thank You all for making Core as beautiful as it is, and I am eager to continue the collaboration towards integrating it properly in Wasabi.

**Comment by luke-jr** [comment]
Neutrino is NOT full verification, and people falsely portraying it as such is a huge part of the problem with Neutrino. It is no safer than other SPV protocols, and less efficient.

I don't see why this means Wasabi will forego full validation. That doesn't logically follow at all.

**Comment by MaxHillebrand** [comment]
Yes @luke-jr, the current way BIP158 is used in Wasabi relies on the server to build and send valid filters - there is no verification, and trust in the server. [this is the problem that we intended to solve with this PR]. Wasabi uses block filters solely for the privacy benefits, of having a light wallet without leaking the xpub or address to any third party server.

However, we intended to use SegWit v0 only [later SegWit v1 only] block filters generated in the [Wasabi in-built bitcoind](https://github.com/zkSNACKs/WalletWasabi/tree/master/WalletWasabi/Microservices/Binaries/lin64), thus every Wasabi user, who opts in to run bitcoind, can cross reference the server filters, from his own generated ones, thus having full verification.

**Comment by MarcoFalke** [comment]
> even after segwit activation, basic filters are 4.8 GB; compared to 250MB for segwit v0 filters

Segwit usage is higher than 50% right now, and assuming that most wallets have a rather recent birthday, the difference should be a factor of 2 or so (clearly less than 10). Further, assuming that taproot is activated some day, it will likely show a similar or higher usage pattern. With that in mind, promising users storage requirements that can not hold over the long term anyway and converge to those of the all-including basic BIP158 filter, seems slightly misleading.

Finally, network bandwidth shouldn't matter when users feed filters from their local bitcoind instance.

**Comment by sipa** [comment]
If it's for local RPC usage, why is a filter needed at all? You can just fetch the block, and check it for matching transactions like a pre-BIP37 wallet would. If that results in any transactions that were missing from the remotely-provided filter, you know the filter is missing things.

**Comment by luke-jr** [comment]
> Yes @luke-jr, the current way BIP158 is used in Wasabi relies on the server to build and send valid filters - there is no verification, and trust in the server. [this is the problem that we intended to solve with this PR].

This PR cannot solve it.

> Wasabi uses block filters solely for the privacy benefits, of having a light wallet without leaking the xpub or address to any third party server.

There are no privacy benefits when you're using your own node.

>However, we intended to use SegWit v0 only [later SegWit v1 only] block filters generated in the Wasabi in-built bitcoind, thus every Wasabi user, who opts in to run bitcoind, can cross reference the server filters, from his own generated ones, thus having full verification.

There is no need for a third-party server filter. Just use your own.

**Comment by nopara73** [comment]
Let me clarify a few things.  

Wasabi client uses P2WPKH only filters those are built on and fetched from the Wasabi backend. We want to make Wasabi a hybrid full node, which we've progressed towards gradually, to the point that only the validation function is needed to be implemented. The idea is to serve the exact same filters from the server to the clients, so the client can just verify the correct filters are provided by the server until its own local Bitcoin Core node catches up. Wasabi wallets aren't running all the time, they are closed and reopened constantly, but ultimately what really matters is that eventually the validation happens, the when is less important, IMO.

Porting our own filter building implementation to the client or going through every block for every wallet is no-go as it has poor performance, because going through every block with RPC is a slow process. You may take it as a given as you are working on Bitcoin Core, but the performance of Bitcoin Core is impossible to even come close to for any project.

What just came to my mind, the most important limitation for most performance issues while communicating with Core is that we cannot just pull out the input value-script pars when getting a transaction from Core. Would that be something sensible to start working on? Like a getblock rpc request where more information on the inputs are added to the tx?

**Comment by luke-jr** [comment]
>Like a getblock rpc request where more information on the inputs are added to the tx?

Like #16083 ?

File: src/blockfilter.cpp (hunk 9 above)
[the remark this answers is not in the document]
Comment (Sjors) [reply src/blockfilter.cpp]: Aargh, I totally misread that code :-)

**Comment by Sjors** [comment]
It may be worth studying c-lightning; they scan all blocks for lightning transactions, no filters needed, and in my experience it's fast. You could wrap some of their C code if it looks too magical :-)

**Comment by nopara73** [comment]
> Like #16083 ? (It's already in Knots, so you could even use it immediately!)

image

**Comment by harding** [comment]
@MaxHillebrand 

> Basic filters are unfortunately not feasible for Wasabi [even after segwit activation, basic filters are 4.8 GB; compared to 250MB for segwit v0 filters].

As @MarcoFalke noted above, this conclusion seems unlikely, so I tested the size of the current "basic" filters myself from segwit activation to present and I get a size of 2.6 GB.

    $ for i in `seq 481824 620831` ; do bitcoin-cli getblockhash $i ; done | while read hash ; do bitcoin-cli getblockfilter $hash ; done | jq -r .filter | xxd -r -p- | wc -c
    2628728815

Beyond that, I'm confused by this disk space argument in the first place.  If you're getting the filter for block x from your local node, don't you also expect to be getting the contents of block x from the same node?  Blocks are ~33x larger than a full basic filter for that block, so the extra few kilobytes disk space per filter (for present usage patterns) seems particularly irrelevant.

@nopara73 

> The idea is to serve the exact same filters from the server to the clients, so the client can just verify the correct filters are provided by the server until its own local Bitcoin Core node catches up.

I understand the idea is for the Wasabi application to get identical filters from either Wasabi's server or the local Bitcoin Core node, but I want to make sure you're aware that, as long as the filter encoding is the same, the Wasabi application should be able to use the basic filters produced by Core without any changes to Wasabi's application logic.  That's because the application doesn't care whether the filter was constructed out of just P2WPKH transactions or all transactions---it's just looking for potential output/input scriptPubKey matches to its wallet.

So, unless I'm misunderstanding something, the Wasabi application should be able to transition seamlessly between using the P2WPKH-only filters served by the Wasabi server to basic (everything) filters served by the local Bitcoin Core node after it has caught up.

**Comment by dangershony** [comment]
@harding yeah this was incorrect the 4.8GB refers to the entire basic filter data, still core has no way to store locally only basic filters form segwit activation. (also note that 2.6 GB is probably too much if we can avoid it)

> the Wasabi application should be able to use the basic filters produced by Core without any changes to Wasabi's application logic.  

Wasabi compresses the script on the clients before trying to match it with a filter (or more correctly wasabi builds the filter with compressed scripts) so legacy clients won't work with basic filters from core.

**Comment by harding** [comment]
> 2.6 GB is probably too much if we can avoid it

2.6 is roughly 50% the minimum amount of data Bitcoin Core needs to store anyway (chainstate plus minimum prune depth) or the rough equivalent of 10 days worth of blocks and undo data at current utilization levels, so 2.6 GB seems to me to be a pretty small additional requirement on top of using Bitcoin Core in the first place.

And, as mentioned elsewhere in this thread, you can use the wallet birthday to prune filters from blocks before the wallet was created (though I don't think we currently support filter pruning).

> wasabi builds the filter with compressed scripts) so legacy clients won't work with basic filters from core.

I'm confused.  What are "compressed scripts"?  I skimmed this PR---it is a nice example of a minimalistic change, thanks!---and it seemed to me that it builds both the basic and v0 filters the same way, except that the v0 filter skips past any transactions with non-v0 parts:

```
if (!script.IsWitnessProgram(witnessversion, witnessprogram)) 
if (witnessversion != witness_version) 
if (!(witnessversion == 0 && (witnessprogram.size() == WITNESS_V0_KEYHASH_SIZE || witnessprogram.size() == WITNESS_V0_SCRIPTHASH_SIZE))) continue; // specific v0 checks
```

Unless I'm missing something, that should mean that any Wasabi clients compatible with this PR would also automatically be compatible with basic filters with no changes needed (except maybe to deal with the break in the chain of filter headers).

**Comment by dangershony** [comment]
You are right, this filter will sill not be compatible with wasabi (and had we gone down this path would mean we have to support two api endpoints after the upgrade) there are a few narratives in parallel here (fetch filters from local core node, fetch filters from wasabi server, maintain backwards compatibility, least disruption to current wasabi design etc...) this will explain the confusion.

>  so 2.6 GB seems to me to be a pretty small additional requirement on top of using Bitcoin Core in the first place. 

Each client (not using a local core node) will fetch 2.6GB of data form the server (wasabi stores all filters locally) if we can prevent this its preferable.

Right new it seems we probably take the approach of using Knots and constructing the same filters wasabi currently uses from a Knots node (that can serve the spent ScriptPubKey)
https://github.com/zkSNACKs/WalletWasabi/issues/3208#issuecomment-595768764

FYI compressed scripts to bytes https://github.com/zkSNACKs/WalletWasabi/blob/master/WalletWasabi/Blockchain/BlockFilters/IndexBuilderService.cs#L264

**Comment by harding** [comment]
> Each client (not using a local core node) will fetch 2.6GB of data form the server (wasabi stores all filters locally) if we can prevent this its preferable.

Yes, this can be prevented, that's what I've been trying to say but I must not be communicating effectively here, so let me start from first principles (my appologies for repeating stuff you probably already know).

The process of creating a BIP158-style filter starts with a block that's full of transactions which each spend one or more UTXOs and which each create one or more new UTXOs.  For BIP158, scriptPubKeys for all spent UTXOs and created UTXOs (excepting `OP_RETURN` outputs) are hashed and then put into a sorted list:

    00000000
    11111111
    22222222
    33333333
    
The list is then golumb coded to minimize its size to produce the final BIP158 basic filter.  The client then takes a list of its scriptPubKeys, hashes them the same way, and compares each of them to the entries in the filter to see if there's any matches.  E.g., the wallet matches on `22222222` and so goes and downloads the block to find the possibly-relevant transaction.

Now, as proposed in this PR, you can create a filter using a subset of the original data (e.g. only P2WPKH):

    11111111
    22222222

If you use the same hashing and golumb coding, then any client which can process the subset filter above can also process the original full set filter to find its scriptPubKey `22222222`.

This means Wasabi's server can generate and serve the small subset filter (~250 MB currently you say) while it's also possible for clients to use their local node's full filter (~2.6 GB post-segwit activation).  This is my point: you don't have to choose between a 250 MB and a 2.6 GB filter---you can use them both in different contexts with no special logic in the client.

> Right new it seems we probably take the approach of using Knots and constructing the same filters wasabi currently uses from a Knots node (that can serve the spent ScriptPubKey)

If you do indeed have a custom GCS algorithm, that does seem like it might be the simplest path forward for you, but it seems a shame to me that choosing that path would remove the possibility of compatibility with the most popular and best reviewed full node implementation (as well as all other software that's implemented standard BIP158).

> FYI compressed scripts to bytes https://github.com/zkSNACKs/WalletWasabi/blob/master/WalletWasabi/Blockchain/BlockFilters/IndexBuilderService.cs#L264

That looks like it's just doing the GCS compression; I'm not sure how your code differs from the same operation as described in BIP158.

**Pull request closed without merging**